### PR TITLE
Adding win_nssm module

### PR DESCRIPTION
**Pull request by georgefrank**
This module provides an interface to the "Non-Sucking Service Manager" program available for windows. The module assumes "nssm" is available via command line (it can be installed via chocolatey).

### CODE REVIEW

**Comment by jhawkesworth** [comment]
Thanks for this, this is going to be really useful - have had good experiences with nssm already so will be great to have a module for it.  

First thought on the PR is to mention that you need to install nssm actually in the documentation .py file.  I think there might even be a 'requirements' section in the documentation options which would probably be the place to mention it - have a look at some core modules.

Hope to test soon although not sure just when at the moment.

**Comment by gregdek** [comment]
Thanks for submitting this new module.

To help facilitate reviews, we’ve broadened the number of people who can approve modules for inclusion into the Extras repository. The list of official reviewers can be found here:

https://github.com/ansible/ansible-modules-extras/blob/devel/REVIEWERS.md 

Our new policy is that if a new module is reviewed and approved by at least two official module reviewers, and receives no -1 votes, the module will be approved for inclusion.  We will be asking the community of reviewers to take a look at these modules on a regular basis.

To ensure that your module has the best chance of being approved, please double-check that you adhere to the Ansible module guidelines: http://docs.ansible.com/developing_modules.html#module-checklist

cc: @trondhindenes @petemounce @elventear @smadam813 @jhawkesworth @angstwad @sivel @chrishoffman @cchurch

**Comment by chrishoffman** [comment]
+1

**Comment by smadam813** [comment]
Looks good!  +1

File: windows/win_nssm.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  app_parameters:
+    description:
+      - Parameters to be passed to the application when it starts
+    required: false
+    default: null
+    aliases: []
+author: "Adam Keech (@smadam813), George Frank (@georgefrank)"
+'''
+
+# TODO:
```
Comment (petemounce) [inline windows/win_nssm.py]: Is this comment relevant to this module?

File: windows/win_nssm.ps1 [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+
+Function Service-Exists
+{
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory=$true)]
+        [string]$name
+    )
+
+    return ,[bool](Get-Service "$name" -ErrorAction SilentlyContinue)
```
Comment (petemounce) [inline windows/win_nssm.ps1]: Leading comma typo...?

File: windows/win_nssm.py (hunk 1 above)
Comment (georgefrank) [inline windows/win_nssm.py]: Correct - it's not relevant. I'll remove that.

File: windows/win_nssm.ps1 (hunk 2 above)
Comment (georgefrank) [inline windows/win_nssm.ps1]: This one isn't needed, so I'll remove it (it was just habit!). The leading comma does, however, have a use when multiple items are returned (more info: http://stackoverflow.com/questions/25279112/returning-a-liststring-using-a-comma-vs-not-using-a-comma)

Comment (petemounce) [inline windows/win_nssm.ps1]: Ah, thank you - every day's a school day!

File: windows/win_nssm.ps1 [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory=$true)]
+        [string]$name
+    )
+
+    if (Service-Exists -name $name)
+    {
+        $cmd = "nssm stop ""$name"""
+        $results = invoke-expression $cmd
```
Comment (petemounce) [inline windows/win_nssm.ps1:111]: Can this fail/throw?

File: windows/win_nssm.ps1 [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+    param(
+        [Parameter(Mandatory=$true)]
+        [string]$name,
+        [Parameter(Mandatory=$true)]
+        [string]$application
+    )
+
+    #note: the application name must look like the following, if the directory includes spaces:
+    # nssm install service "c:\Program Files\app.exe" """C:\Path with spaces"""
+    #see https://git.nssm.cc/?p=nssm/nssm.git;a=commit;h=0b386fc1984ab74ee59b7bed14b7e8f57212c22b for more info
```
Comment (petemounce) [inline windows/win_nssm.ps1]: This would be better in the `py` doc, so it appears in the documentation.

File: windows/win_nssm.ps1 [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+
+
+    if (!$application)
+    {
+        Throw "Error installing service ""$name"". No application was supplied."
+    }
+
+    if (!(Service-Exists -name $name))
```
Comment (petemounce) [inline windows/win_nssm.ps1:142]: Currently, `Service-Exists` returns a collection with an element in it. Does the `!` operator do the intended thing there, if `@(true)` comes back? Unfamiliar with the operator unwrapping as I discovered from earlier reply.

File: windows/win_nssm.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    - nssm is a service helper which doesn't suck. See https://nssm.cc/ for more information.
+requirements:
+  - "nssm >= 2.24.0 # (install via win_chocolatey) win_chocolatey: name=nssm"
+options:
+  name:
+    description:
+      - Name of the service to operate on
+    required: true
+    default: null
+    aliases: []
```
Comment (petemounce) [inline windows/win_nssm.py]: I think an empty alias list can be omitted.

File: windows/win_nssm.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+description:
+    - nssm is a service helper which doesn't suck. See https://nssm.cc/ for more information.
+requirements:
+  - "nssm >= 2.24.0 # (install via win_chocolatey) win_chocolatey: name=nssm"
+options:
+  name:
+    description:
+      - Name of the service to operate on
+    required: true
+    default: null
```
Comment (petemounce) [inline windows/win_nssm.py]: I think if required is true, default is unnecessary to specify.

File: windows/win_nssm.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+      - started
+      - stopped
+      - restarted
+      - absent
+    default: started
+    aliases: []
+
+  application:
+    description:
+      - The application binary to run as a service
```
Comment (petemounce) [inline windows/win_nssm.py:54]: Since required is false and default is null, it would be nice to expand the description to mention that is a parameter to use when the service is already installed (I assume)

File: windows/win_nssm.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    required: false
+    default: null
+    aliases: []
+  
+  stdout_file:
+    description:
+      - Path to receive output
```
Comment (petemounce) [inline windows/win_nssm.py:64]: A note about what happens with relative/absolute path would be nice - are relative paths relative to the application binary, or the nssm binary, or...?

**Comment by petemounce** [comment]
@georgefrank thanks very much for the contribution; I'm +1 for it becoming merged in terms of code review, but am not in a position to actually test it at the moment.

File: windows/win_nssm.ps1 (hunk 5 above)
Comment (georgefrank) [inline windows/win_nssm.ps1:142]: `!@($true)` yields the same result as `!$true` (also, `@($true)` and `$true` are equivalent as well)

File: windows/win_nssm.ps1 (hunk 4 above)
Comment (georgefrank) [inline windows/win_nssm.ps1]: Completely agree! I'll move it there.

**Comment by georgefrank** [comment]
Thanks for all the feedback, @petemounce! I'm not used to writing formal documentation, so all of the feedback there is especially appreciated. I'll fix things up and re-submit soon(:tm:).

File: windows/win_nssm.py (hunk 6 above)
Comment (georgefrank) [inline windows/win_nssm.py]: There are a bunch of empty alias collections, so I'll remove all of them.

File: windows/win_nssm.py (hunk 8 above)
Comment (georgefrank) [inline windows/win_nssm.py:54]: It's needed any time the service needs to be installed, which may happen if you specify any state other than "absent". This clearly needs a better description!

File: windows/win_nssm.ps1 (hunk 3 above)
Comment (georgefrank) [inline windows/win_nssm.ps1:111]: Can you elaborate on what you're asking here? I'm not sure I understand the intent of the question.

Comment (petemounce) [inline windows/win_nssm.ps1:111]: In the statement below you check and handle an error code, but not here. I
don't know nssm's behaviour. Is it not necessary to check the exit code
here?

Sorry for not being clear.

##

File: windows/win_nssm.ps1 (hunk 5 above)
Comment (petemounce) [inline windows/win_nssm.ps1:142]: Cool

**Comment by petemounce** [comment]
@georgefrank lgtm pending an answer on the exit-code check line 115. Otherwise, +1 and thanks very much for contributing!

File: windows/win_nssm.ps1 (hunk 3 above)
Comment (georgefrank) [inline windows/win_nssm.ps1:111]: The stop command can fail, yes; however, we're only interested in whether the service can be removed (since we're already sure it exists, per the `Service-Exists` check). Removing a service using nssm only works well if the service is stopped. Here are the scenarios that can play out:
1. The service is stopped. In this case, `nssm stop` will fail, but `nssm remove` will succeed, so we don't want to do anything about the `nssm stop` failure.
2. The service is hanging and won't respond. In this case, `nssm stop` will fail, and also `nssm remove` will fail, which is caught below.
3. The service is responding, but is not currently stopped. In this case, `nssm stop` will succeed, and so will `nssm remove`.

So this command is fine without an error check.

Comment (petemounce) [inline windows/win_nssm.ps1:111]: Gotcha

**Comment by petemounce** [comment]
Ship it!

**Comment by robynbergeron** [comment]
@georgefrank thanks for the PR; @chrishoffman, @smadam813, @jhawkesworth, @petemounce, thanks for the reviews and helping work through this. :)

Marking for inclusion; cc: @ansible/core team.
